File: src/test/java/com/trilead/ssh2/DropbearCompatibilityTest.java
package com.trilead.ssh2;

import org.apache.commons.io.IOUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.output.Slf4jLogConsumer;
import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy;
import org.testcontainers.images.builder.ImageFromDockerfile;

import java.io.IOException;
import java.util.function.Consumer;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

/**
 * Integration tests against Dropbear.
 *
 * @author Kenny Root
 */
public class DropbearCompatibilityTest {
	private static final Logger logger = LoggerFactory.getLogger(DropbearCompatibilityTest.class.getSimpleName());
	private static final Slf4jLogConsumer logConsumer = new Slf4jLogConsumer(logger).withPrefix("DOCKER");

	@Rule
	public SshLogger sshLogger = new SshLogger(logger);

	private static final String OPTIONS_ENV = "OPTIONS";
	private static final String USERNAME = "testuser";
	private static final String PASSWORD = "testtest123";

	private static ImageFromDockerfile baseImage = new ImageFromDockerfile("dropbear-server", false)
			.withFileFromClasspath(".", "dropbear-server");

	static {
		for (String key : PubkeyConstants.KEY_NAMES) {
			baseImage.withFileFromClasspath(key, "com/trilead/ssh2/crypto/" + key);
		}
	}

	@ClassRule
	public static GenericContainer server = getBaseContainer();

	@NotNull
	@Contract("_ -> new")
	private Connection withServer(@NotNull GenericContainer container) {
		return new Connection(container.getContainerIpAddress(), container.getMappedPort(22));
	}

	private static GenericContainer getBaseContainer() {
		return new GenericContainer(baseImage)
				.withExposedPorts(22)
				.withLogConsumer(logConsumer)
				.waitingFor(new LogMessageWaitStrategy()
						.withRegEx(".*Not backgrounding.*\\s"));
	}

	private ConnectionInfo assertCanPasswordAuthenticate(GenericContainer server) throws IOException {
		return assertCanPasswordAuthenticate(server, null);
	}

	private ConnectionInfo assertCanPasswordAuthenticate(GenericContainer server, Consumer<Connection> setupFunc) throws IOException {
		try (Connection c = withServer(server)) {
			if (setupFunc != null) {
				setupFunc.accept(c);
			}
			c.connect();
			assertThat(c.authenticateWithPassword(USERNAME, PASSWORD), is(true));
			try (Session s = c.openSession()) {
				s.ping();
			}
			return c.getConnectionInfo();
		}
	}

	private ConnectionInfo connectToServerWithOptions(@NotNull String options) throws IOException {
		return connectToServerWithOptions(options, null);
	}

	private ConnectionInfo connectToServer(@Nullable Consumer<Connection> setupFunc) throws IOException {
		return assertCanPasswordAuthenticate(server, setupFunc);
	}

	private ConnectionInfo connectToServerWithOptions(@NotNull String options, @Nullable Consumer<Connection> setupFunc) throws IOException {
		try (GenericContainer customServer = getBaseContainer().withEnv(OPTIONS_ENV, options)) {
			customServer.start();
			return assertCanPasswordAuthenticate(customServer, setupFunc);
		}
	}

	private void assertCanConnectToServerThatHasKeyType(@NotNull String keyPath, String keyType) throws IOException {
		ConnectionInfo info = connectToServerWithOptions("-r " + keyPath);
		assertThat(info.serverHostKeyAlgorithm, is(keyType));
	}

	private void canConnectWithPubkey(String keyFilename) throws Exception {
		char[] keyChars = IOUtils.toCharArray(getClass().getResourceAsStream("crypto/" + keyFilename), "UTF-8");

		try (GenericContainer server = getBaseContainer()) {
			server.start();
			try (Connection connection = withServer(server)) {
				connection.connect();
				assertThat(connection.authenticateWithPublicKey(USERNAME, keyChars, ""), is(true));
				try (Session session = connection.openSession()) {
					session.ping();
				}
			}
		}
	}

	@Test
	public void canConnectWithEcdsa256() throws Exception {
		canConnectWithPubkey("ecdsa-nistp256-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithEcdsa384() throws Exception {
		canConnectWithPubkey("ecdsa-nistp384-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithEcdsa521() throws Exception {
		canConnectWithPubkey("ecdsa-nistp521-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithRsa() throws Exception {
		canConnectWithPubkey("rsa-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithPassword() throws Exception {
		try (GenericContainer server = getBaseContainer()) {
			server.start();
			assertCanPasswordAuthenticate(server);
		}
	}

	@Test
	public void wrongPasswordFails() throws Exception {
		try (GenericContainer server = getBaseContainer()) {
			server.start();
			try (Connection c = withServer(server)) {
				c.connect();
				assertThat(c.authenticateWithPassword(USERNAME, "wrongpassword"), is(false));
			}
		}
	}

	@Test
	public void connectToRsaHost() throws Exception {
		assertCanConnectToServerThatHasKeyType("/etc/dropbear/dropbear_rsa_host_key", "rsa-sha2-256");
	}

	@Test
	public void connectToEcdsaHost() throws Exception {
		assertCanConnectToServerThatHasKeyType("/etc/dropbear/dropbear_ecdsa_host_key", "ecdsa-sha2-nistp256");
	}

	private void assertCanConnectToServerWithKex(@NotNull String kexType) throws IOException {
		ConnectionInfo info = connectToServer(
				c -> c.setKeyExchangeAlgorithms(new String[]{kexType}));
		assertThat(info.keyExchangeAlgorithm, is(kexType));
	}

	@Test
	public void canConnectWithKexCurve25519LibsshOrg() throws Exception {
		assertCanConnectToServerWithKex("curve25519-sha256@libssh.org");
	}

	@Test
	public void canConnectWithKexCurve25519() throws Exception {
		assertCanConnectToServerWithKex("curve25519-sha256");
	}

	@Test
	public void canConnectWithKexDHGroup14() throws Exception {
		assertCanConnectToServerWithKex("diffie-hellman-group14-sha1");
	}

	@Test
	public void canConnectWithKexDHGroup14Sha256() throws Exception {
		assertCanConnectToServerWithKex("diffie-hellman-group14-sha256");
	}

	@Test
	public void canConnectWithKexEcdhSha2Nistp256() throws Exception {
		assertCanConnectToServerWithKex("ecdh-sha2-nistp256");
	}

	@Test
	public void canConnectWithKexEcdhSha2Nistp384() throws Exception {
		assertCanConnectToServerWithKex("ecdh-sha2-nistp384");
	}

	@Test
	public void canConnectWithKexEcdhSha2Nistp521() throws Exception {
		assertCanConnectToServerWithKex("ecdh-sha2-nistp521");
	}

	private void setCiphers(Connection c, String cipher) {
		c.setClient2ServerCiphers(new String[]{cipher});
		c.setServer2ClientCiphers(new String[]{cipher});
	}

	private void assertCanConnectToServerWithCipher(@NotNull String cipher) throws IOException {
		ConnectionInfo info = connectToServer(c -> setCiphers(c, cipher));
		assertThat(info.clientToServerCryptoAlgorithm, is(cipher));
		assertThat(info.serverToClientCryptoAlgorithm, is(cipher));
	}

	@Test
	public void canConnectWithCipherAes128Ctr() throws Exception {
		assertCanConnectToServerWithCipher("aes128-ctr");
	}

	@Test
	public void canConnectWithCipherAes256Ctr() throws Exception {
		assertCanConnectToServerWithCipher("aes256-ctr");
	}

	private void setMac(Connection c, String mac) {
		c.setClient2ServerMACs(new String[]{mac});
		c.setServer2ClientMACs(new String[]{mac});
	}

	private void assertCanConnectToServerWithMac(@NotNull String mac) throws IOException {
		ConnectionInfo info = connectToServer(c -> setMac(c, mac));
		assertThat(info.clientToServerMACAlgorithm, is(mac));
		assertThat(info.serverToClientMACAlgorithm, is(mac));
	}

	@Test
	public void canConnectWithMacHmacSha1() throws Exception {
		assertCanConnectToServerWithMac("hmac-sha1");
	}

	@Test
	public void canConnectWithMacHmacSha2_256() throws Exception {
		assertCanConnectToServerWithMac("hmac-sha2-256");
	}

	@Test
	public void canConnectWithCompression() throws Exception {
		try (GenericContainer customServer = getBaseContainer()) {
			customServer.start();
			try (Connection c = withServer(customServer)) {
				c.setCompression(true);
				c.connect();
				assertThat(c.authenticateWithPassword(USERNAME, PASSWORD), is(true));
				try (Session s = c.openSession()) {
					s.ping();
				}

				ConnectionInfo info = c.getConnectionInfo();
				assertThat(info.clientToServerCompressionAlgorithm, is("zlib@openssh.com"));
				assertThat(info.serverToClientCompressionAlgorithm, is("zlib@openssh.com"));
			}
		}
	}

	private void canConnectWithHostKeyAlgorithm(String keyPath, String hostKeyAlgorithm) throws Exception {
		ConnectionInfo info = connectToServerWithOptions("-r " + keyPath, c -> c.setServerHostKeyAlgorithms(new String[]{hostKeyAlgorithm}));
		assertThat(info.serverHostKeyAlgorithm, is(hostKeyAlgorithm));
	}

	@Test
	public void canConnectToHostWithHostKeySshRsa() throws Exception {
		canConnectWithHostKeyAlgorithm("/etc/dropbear/dropbear_rsa_host_key", "ssh-rsa");
	}
}

File: sshlib/src/test/kotlin/org/connectbot/sshlib/client/DropbearCompatibilityTest.kt
/*
 * ConnectBot SSH Library
 * Copyright 2025-2026 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.sshlib.client

import org.connectbot.sshlib.HostKeyVerifier
import org.connectbot.sshlib.PublicKey
import org.connectbot.sshlib.SshClientConfig
import org.connectbot.sshlib.SshException
import org.connectbot.sshlib.blocking.BlockingSshClient
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import org.slf4j.LoggerFactory
import org.testcontainers.containers.GenericContainer
import org.testcontainers.containers.output.Slf4jLogConsumer
import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy
import org.testcontainers.images.builder.ImageFromDockerfile
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@Testcontainers
class DropbearCompatibilityTest {

    companion object {
        private val logger = LoggerFactory.getLogger(DropbearCompatibilityTest::class.java)
        private val logConsumer = Slf4jLogConsumer(logger).withPrefix("DOCKER")

        private const val OPTIONS_ENV = "OPTIONS"
        private const val USERNAME = "testuser"
        private const val PASSWORD = "testpass"

        private val PUBLIC_KEY_NAMES = arrayOf(
            "ed25519_unencrypted.pub",
            "ecdsa256_unencrypted.pub",
            "ecdsa384_unencrypted.pub",
            "ecdsa521_unencrypted.pub",
            "rsa_unencrypted.pub",
        )

        private val baseImage = ImageFromDockerfile("dropbear-server-test", false)
            .withFileFromClasspath(".", "dropbear-server").also { image ->
                for (key in PUBLIC_KEY_NAMES) {
                    image.withFileFromClasspath(key, "keys/$key")
                }
            }

        @Container
        @JvmStatic
        val server: GenericContainer<*> = createBaseContainer()

        @Suppress("resource")
        private fun createBaseContainer(): GenericContainer<*> = GenericContainer(baseImage)
            .withExposedPorts(22)
            .withLogConsumer(logConsumer)
            .waitingFor(LogMessageWaitStrategy().withRegEx(".*Not backgrounding.*\\s"))

        @JvmStatic
        fun encryptionAlgorithms() = listOf(
            "aes128-ctr",
            "aes256-ctr",
            "chacha20-poly1305@openssh.com",
        )

        @JvmStatic
        fun kexAlgorithms() = listOf(
            "mlkem768x25519-sha256",
            "curve25519-sha256",
            "ecdh-sha2-nistp256",
            "ecdh-sha2-nistp384",
            "ecdh-sha2-nistp521",
            "diffie-hellman-group14-sha256",
        )

        @JvmStatic
        fun macAlgorithms() = listOf(
            "hmac-sha2-256",
        )

        @JvmStatic
        fun pubkeyFiles() = listOf(
            "ed25519_unencrypted",
            "ecdsa256_unencrypted",
            "ecdsa384_unencrypted",
            "ecdsa521_unencrypted",
        )

        @JvmStatic
        fun hostKeyConfigs() = listOf(
            HostKeyConfig("/etc/dropbear/dropbear_ed25519_host_key", "ssh-ed25519"),
            HostKeyConfig("/etc/dropbear/dropbear_ecdsa_host_key", "ecdsa-sha2-nistp256"),
            HostKeyConfig("/etc/dropbear/dropbear_rsa_host_key", "rsa-sha2-256"),
        )

        data class HostKeyConfig(val path: String, val algorithm: String) {
            override fun toString(): String = algorithm
        }
    }

    private val acceptAllVerifier = object : HostKeyVerifier {
        override suspend fun verify(key: PublicKey): Boolean = true
    }

    private fun connectToServer(configure: (SshClientConfig.Builder.() -> Unit)? = null): BlockingSshClient {
        val host = server.host
        val port = server.getMappedPort(22)

        val config = SshClientConfig {
            this.host = host
            this.port = port
            this.hostKeyVerifier = acceptAllVerifier
            configure?.invoke(this)
        }
        val client = BlockingSshClient(config)
        client.connect()
        return client
    }

    @Suppress("resource")
    private fun authenticateWithOptions(
        options: String,
        configure: (SshClientConfig.Builder.() -> Unit)? = null,
    ) {
        val customServer = createBaseContainer().withEnv(OPTIONS_ENV, options)
        customServer.start()
        try {
            val host = customServer.host
            val port = customServer.getMappedPort(22)
            val config = SshClientConfig {
                this.host = host
                this.port = port
                this.hostKeyVerifier = acceptAllVerifier
                configure?.invoke(this)
            }
            val client = BlockingSshClient(config)
            try {
                client.connect()
                client.authenticatePassword(USERNAME, PASSWORD)
            } finally {
                client.disconnect()
            }
        } finally {
            customServer.stop()
        }
    }

    private fun authenticateWithPassword(configure: (SshClientConfig.Builder.() -> Unit)? = null) {
        val client = connectToServer(configure)
        try {
            client.authenticatePassword(USERNAME, PASSWORD)
        } finally {
            client.disconnect()
        }
    }

    @Test
    fun `can connect with password`() {
        authenticateWithPassword()
    }

    @Test
    fun `wrong password fails`() {
        val client = connectToServer()
        try {
            assertThrows<SshException> { client.authenticatePassword(USERNAME, "wrongpassword") }
        } finally {
            client.disconnect()
        }
    }

    @ParameterizedTest(name = "pubkey: {0}")
    @MethodSource("pubkeyFiles")
    fun `can connect with public key`(keyFilename: String) {
        val keyData = javaClass.getResourceAsStream("/keys/$keyFilename")!!
            .bufferedReader().readText()
        val client = connectToServer()
        try {
            client.authenticatePublicKey(USERNAME, keyData)
        } finally {
            client.disconnect()
        }
    }

    @ParameterizedTest(name = "hostkey: {0}")
    @MethodSource("hostKeyConfigs")
    fun `can connect to host with key type`(config: HostKeyConfig) {
        authenticateWithOptions("-r ${config.path}") {
            hostKeyAlgorithms = config.algorithm
        }
    }

    @ParameterizedTest(name = "kex: {0}")
    @MethodSource("kexAlgorithms")
    fun `can connect with kex algorithm`(algorithm: String) {
        authenticateWithPassword {
            kexAlgorithms = "$algorithm,kex-strict-c-v00@openssh.com"
        }
    }

    @ParameterizedTest(name = "encryption: {0}")
    @MethodSource("encryptionAlgorithms")
    fun `can connect with encryption algorithm`(algorithm: String) {
        authenticateWithPassword {
            encryptionAlgorithms = algorithm
            macAlgorithms = "hmac-sha2-256"
        }
    }

    @ParameterizedTest(name = "mac: {0}")
    @MethodSource("macAlgorithms")
    fun `can connect with mac algorithm`(algorithm: String) {
        authenticateWithPassword {
            encryptionAlgorithms = "aes128-ctr"
            macAlgorithms = algorithm
        }
    }
}
